File: sdk/node/fluxcast-fcdp.mjs
/** Strict, dependency-free FCDP v0.1 framing SDK for Node.js 20+. */
export const HEADER_LEN = 37;
export function crc16(bytes) { let value = 0xffff; for (const byte of bytes) { value ^= byte << 8; for (let bit = 0; bit < 8; bit += 1) value = value & 0x8000 ? ((value << 1) ^ 0x1021) & 0xffff : (value << 1) & 0xffff; } return value; }
export function encode(header, payload) {
  if (!Buffer.isBuffer(payload)) payload = Buffer.from(payload);
  if (payload.length + HEADER_LEN > 1200) throw new Error("FCDP datagram exceeds 1200-byte budget");
  if (header.priority > 3 || header.fragmentIndex >= header.fragmentCount) throw new Error("invalid FCDP fragment range");
  const raw = Buffer.alloc(35); raw.write("FC", 0, "ascii"); raw[2] = 1; raw[3] = header.packetType ?? 3; raw[4] = header.flags ?? 0;
  raw.writeBigUInt64BE(BigInt(header.sessionId ?? 1), 6); raw.writeUInt16BE(header.streamId ?? 1, 14); raw.writeUInt16BE(header.epoch ?? 0, 16); raw.writeUInt32BE(header.sequence ?? 1, 18); raw.writeUInt32BE(header.frameId ?? 1, 22); raw.writeUInt16BE(header.fragmentIndex ?? 0, 26); raw.writeUInt16BE(header.fragmentCount ?? 1, 28); raw[30] = header.priority ?? 0; raw.writeUInt16BE(header.deadlineMs ?? 1000, 31); raw.writeUInt16BE(payload.length, 33);
  const checksum = Buffer.alloc(2); checksum.writeUInt16BE(crc16(raw)); return Buffer.concat([raw, checksum, payload]);
}
export function decode(packet) { if (!Buffer.isBuffer(packet)) packet = Buffer.from(packet); if (packet.length < HEADER_LEN || packet.subarray(0, 2).toString("ascii") !== "FC" || packet[2] !== 1 || crc16(packet.subarray(0, 35)) !== packet.readUInt16BE(35)) throw new Error("invalid FCDP header"); const length = packet.readUInt16BE(33); const payload = packet.subarray(37); if (payload.length !== length) throw new Error("FCDP payload length mismatch"); return { header: { packetType: packet[3], flags: packet[4], sessionId: packet.readBigUInt64BE(6), streamId: packet.readUInt16BE(14), epoch: packet.readUInt16BE(16), sequence: packet.readUInt32BE(18), frameId: packet.readUInt32BE(22), fragmentIndex: packet.readUInt16BE(26), fragmentCount: packet.readUInt16BE(28), priority: packet[30], deadlineMs: packet.readUInt16BE(31) }, payload }; }

File: sdk/kotlin/src/main/kotlin/FluxCastFcdp.kt
package fluxcast

import java.nio.ByteBuffer
import java.nio.ByteOrder

data class FcdpHeader(val packetType: Int=3,val flags:Int=0,val sessionId:Long=1,val streamId:Int=1,val epoch:Int=0,val sequence:Long=1,val frameId:Long=1,val fragmentIndex:Int=0,val fragmentCount:Int=1,val priority:Int=0,val deadlineMs:Int=1000)
object FluxCastFcdp {
 const val HEADER_LEN=37; const val MAX_DATAGRAM=1200
 private fun crc16(data:ByteArray):Int { var v=0xffff; for(x in data){v=v xor ((x.toInt()and 255)shl 8); repeat(8){v=if(v and 0x8000!=0)((v shl 1)xor 0x1021)and 0xffff else (v shl 1)and 0xffff}};return v }
 fun encode(h:FcdpHeader,payload:ByteArray):ByteArray { require(payload.size+HEADER_LEN<=MAX_DATAGRAM&&h.priority in 0..3&&h.fragmentCount>0&&h.fragmentIndex in 0 until h.fragmentCount); val p=ByteBuffer.allocate(HEADER_LEN+payload.size).order(ByteOrder.BIG_ENDIAN);p.put(byteArrayOf(70,67,1,h.packetType.toByte(),h.flags.toByte(),0));p.putLong(h.sessionId);p.putShort(h.streamId.toShort());p.putShort(h.epoch.toShort());p.putInt(h.sequence.toInt());p.putInt(h.frameId.toInt());p.putShort(h.fragmentIndex.toShort());p.putShort(h.fragmentCount.toShort());p.put(h.priority.toByte());p.putShort(h.deadlineMs.toShort());p.putShort(payload.size.toShort());val raw=p.array().copyOfRange(0,35);p.putShort(crc16(raw).toShort());p.put(payload);return p.array() }
 fun decode(packet:ByteArray):Pair<FcdpHeader,ByteArray> { require(packet.size>=HEADER_LEN&&packet[0].toInt()==70&&packet[1].toInt()==67&&packet[2].toInt()==1);require(crc16(packet.copyOfRange(0,35))==ByteBuffer.wrap(packet,35,2).order(ByteOrder.BIG_ENDIAN).short.toInt()and 0xffff);val b=ByteBuffer.wrap(packet).order(ByteOrder.BIG_ENDIAN);b.position(3);val t=b.get().toInt()and 255;val f=b.get().toInt()and 255;b.get();val h=FcdpHeader(t,f,b.long,b.short.toInt()and 65535,b.short.toInt()and 65535,b.int.toLong()and 0xffffffffL,b.int.toLong()and 0xffffffffL,b.short.toInt()and 65535,b.short.toInt()and 65535,b.get().toInt()and 255,b.short.toInt()and 65535);val n=b.short.toInt()and 65535;require(n==packet.size-HEADER_LEN&&h.priority<=3&&h.fragmentCount>0&&h.fragmentIndex<h.fragmentCount);return h to packet.copyOfRange(HEADER_LEN,packet.size) }
}
